test/core/test_rules.py

```diff
@@ -0,0 +1,93 @@
+from test.helper import check_evaluation, evaluate_value
```

Comment:
This is definitely more of the kind of thing we need - thanks!

In fact we actually could use even lower-level tests. These test at the user level, but we we need a lot more especially for pattern matching is tests of the various routines in more isolation of the rest of the system.

Reply:
The reason of proposing these high level tests is that I am pretty sure that we need to change the low level API. What I am trying to do with this is to check that the changes do not move us away from our current degree of compatibility with WL.

Reply:
Yes, this is why end-to-end tests or integration tests are typically written. 

We _also_  (not _instead_) should be considering smaller unit tests., I agree that things here are likely to change and when that happens unit tests are the most fragile and the first to break.

So I guess here I am thinking that in doing larger changes, then it would be great to start adding unit (transparent-box) unit tests.

Reply:
> What I am trying to do with this is to check that the changes do not move us away from our current degree of compatibility with WL.

Thinking about this more. I think this should be in a separate PR and that can go in now.